### DISCUSSION TITLE
Use a deploy token instead of deploy key when bootstrapping with GitLab

### DISCUSSION THREAD

**Discussion by nagyv** [fluxcd/flux2#3595] [Proposals]
Today, Flux bootstrap (and the Terraform provider docs) uses GitLab Deploy Keys to access GitLab repositories.

GitLab offers GitLab tokens that already work with Flux, and might be [better suited](https://docs.gitlab.com/ee/user/project/deploy_keys/) for Flux.

Were there any consideration behind going with Deploy Keys instead of Deploy Tokens? Would you be open to a PR that switches GitLab bootstrap to Deploy tokens?

**Comment by stefanprodan** [comment]
> Today, Flux bootstrap (and the Terraform provider docs) uses GitLab Deploy Keys to access GitLab repositories.

Bootstrap uses a GitLab PAT to clone the repo locally and push upstream the Flux manifests. After that it generates a SSH key, then uses the SSH public key to set it as a Deploy key using the PAT. I'm not sure I understand how a Deploy token would work, is it a replacement for the PAT? Does it allow write access?

**Reply by nagyv** [reply]
With a Deploy Token the flow would be the following: Bootstrap uses a GitLab PAT to clone the repo locally and push upstream the Flux manifests. After that it create a Deploy token using the PAT and sets it as the secret for the GitRepository

> is it a replacement for the PAT? Does it allow write access?

Deploy tokens are a replacement for deploy keys. They are more flexible and secure.  This link has a quick comparison of the two: https://docs.gitlab.com/ee/user/project/deploy_keys/

**Comment by nagyv** [comment]
One shortcoming of deploy tokens is that they work over HTTP-git only, SSH-git is not supported. Are there any Flux features that rely on a SSH GitRepository source?

**Comment by dosuken123** [comment]
I think the one major difference in CD context between Deploy Key and Deploy Token is permissions. Deploy Key is limited to repository access whereas Deploy Token can allow multiple permissions related to deployments, such as private container registry and package registry. So that users can consolidate multiple access secrets into one.

![2023-02-16_10-36](https://user-images.githubusercontent.com/4432788/219242122-2bdd5dd1-7f31-4f60-a117-e04cfb378260.png)

**Comment by stefanprodan** [comment]
Most users rely on SSH for bootstrap, also we have lots of GitLab users who disabled HTTP/S access from the cluster, and only allow SSH. Given this, I think we can add a `--deploy-token` optional arg to `flux bootstrap gitlab`, when this is set, then bootstrap switches to HTTP/S and sets the token in the secret, in the same manner it does today for `--token-auth`.

**Reply by timofurrer** [reply]
I think this would be a nice addition. One (probably not so popular 🤔 ) use-case for this I know of, apart from what has already been mentioned, is that in some regulated environments you want to tightly control outgoing connections. Relying on Deploy Tokens using HTTPS would omit opening up for outgoing SSH. 

I'd be up for contributing that if it's something you'd considering merging :)

**Reply by nagyv** [reply]
Following-up on Slack:

![Képernyőfotó 2023-02-20 - 17 51 01](https://user-images.githubusercontent.com/126671/220163664-72265860-5c46-48e8-ab2f-4ce37a9079f0.png)

**Comment by nagyv** [comment]
I created https://gitlab.com/gitlab-org/gitlab/-/issues/392605 to track the work at GitLab

**Comment by wvh** [comment]
Hello!

Creating a deploy token (not key) with `--deploy-token-auth` works fine, but is always read-only, i.e. scope `read_repository` and not `write_repository`. Maybe this argument should also observe the `--read-write-key` flag; it's confusing for the token to be read-only if this flag is provided.

We need write access for image update automations.

**Reply by stefanprodan** [reply]
Looks like GitLab Deploy Tokens don't allow write access to Git repositories https://docs.gitlab.com/ee/user/project/deploy_tokens/index.html Maybe @nagyv can confirm this limitation, if that's the case, then we should document in the bootstrap page for GitLab, that using `--deploy-token-auth` is incompatible with Flux image automation.

**Comment by nagyv** [comment]
Stefan is right. Deploy tokens don't support repository writes.

I'll open a doc update pr if nobody beats me to it.

-------- Eredeti üzenet --------
2024. ápr. 23. 16:19-kor, wvh ezt írta:

**Reply by nagyv** [reply]
The https://fluxcd.io/flux/installation/bootstrap/gitlab/#gitlab-personal-account page already has an alert that deploy tokens are read-only.

**Reply by stefanprodan** [reply]
Cool, to solve @wvh issue I think the CLI should error out when both `--deploy-token-auth` and `--read-write-key` are provided.

**Comment by Ferdinanddb** [comment]
Hi @nagyv ,

In your original post, you mention the Flux Terraform provider. While looking at the examples provided for Gitlab, I see that there is nothing regarding Deploy Tokens.

I would like to know if the following way of proceeding is ok in your opinion please:

- Create a Gitlab repo manually dedicated to Flux.

- Create a PAT token manually, this PAT token will be used to authenticate the Gitlab provider to my Gitlab instance. I guess that this PAT token should only be granted permissions on the Flux repo.

- Inside the Terraform code, [use the `gitlab_project_deploy_token` resource](https://registry.terraform.io/providers/gitlabhq/gitlab/latest/docs/resources/project_deploy_token) as a replacement of `gitlab_deploy_key` resource [here](https://github.com/fluxcd/terraform-provider-flux/blob/v1.7.4/examples/gitlab-via-ssh/main.tf#L52) and [here](https://github.com/fluxcd/terraform-provider-flux/blob/v1.7.4/examples/gitlab-via-ssh/providers.tf#L14).

Then the bootstrapping process should work right?

---

Also, one thing that I find a bit dirty here, is the fact that the deploy token value is written in the terraform state file. And I think that the new feature of OpenTofu about[ `ephemeral` resources](https://opentofu.org/blog/help-us-test-opentofu-1-11-0-beta1/#ephemeral-resources--write-only-attributes---beta-release) would be awesome here (would allow to create the k8s secret for deploy token in k8s cluster without having to persist the token), right?

**Reply by nagyv** [reply]
You don't need the PAT token. Flux will authenticate with the deploy token. Otherwise, you are correct.

On ephemeral resources, I don't know, never tried it, but sounds legit. Currently, I prefer using the [Flux Operator](https://fluxcd.control-plane.io/operator/install/), instead of setting up a FluxCD instance directly. I deploy the operator with Terraform, including the first Flux instance. It still helps to avoid terraform state issues as the Flux instance is upgraded by the operator, without causing a dirty state.

**Reply by Ferdinanddb** [reply]
> You don't need the PAT token. Flux will authenticate with the deploy token.

Are you sure that we don't need a PAT token? I am asking because for Flux to be bootstrapped then the token used needs to have write access to the Gitlab repo for setting up the Gitlab repo (one time operations to initialize everything --> pushing all the folders and YAML files related to the cluster we are bootstrapping in the Gitlab repo), right? But once the bootstrap is done then I agree that only the deploy token will be used, as Flux is just pulling from Gitlab (with read access).

Regarding what I just wrote, I assume that the deploy token has only read access to the Gitlab repo even though it might be possible to create a deploy token with write access, because I don't want the Flux resources on k8s to authenticate with a deploy token having write access (since they only need read access). Hope I am clear!

Thanks for letting me know about the Flux Operator, I will have a look!

But I believe that this implies installing the Flux Operator using Helm via Terraform?

I remember that I had difficulties (like cyclic errors or something else) in the past when using the Helm provider in the same Terraform project where the kubernetes cluster is created. To rephrase, I remember having trouble to create in one `plan + apply` action both the k8s cluster and install a Helm chart on it, when starting from scratch. I think it was due to the fact that the Helm provider was trying to interact with the k8s cluster at the plan stage, but the k8s cluster did not yet exist.

- The solution was to have two isolated Terraform projects (one to create the kubernetes cluster, one to install Helm charts on it). But maybe something change regarding that? Or maybe the good practice is to have isolated Terraform projects?

**Reply by nagyv** [reply]
> Are you sure that we don't need a PAT token? I am asking because for Flux to be bootstrapped then the token used needs to have write access to the Gitlab repo for setting up the Gitlab repo (one time operations to initialize everything --> pushing all the folders and YAML files related to the cluster we are bootstrapping in the Gitlab repo), right? But once the bootstrap is done then I agree that only the deploy token will be used, as Flux is just pulling from Gitlab (with read access).

You are correct. I misunderstood you. I thought you meant that you want to store the PAT in the cluster as a Secret for some reason.

> But I believe that this implies installing the Flux Operator using Helm via Terraform?

Correct

**Reply by matheuscscp** [reply]
> But I believe that this implies installing the Flux Operator using Helm via Terraform?

This is our recommendation

**Reply by Ferdinanddb** [reply]
Thank you for your answers!

And regarding the use of the Helm provider with Terraform: do you typically have one Terraform project to create the kubernetes cluster (and other resources), and another isolated Terraform project for the Helm charts to be installed on the kubernetes cluster created by the first Terraform project?

I am asking because I would like to avoid having a Terraform project only to install the Flux operator if that is possible, since all the other Helm charts and k8s resources will be managed via Flux and the Gitlab repo.

**Reply by matheuscscp** [reply]
We address these questions in the D2 Reference Architecture: https://fluxcd.control-plane.io/guides/d2-architecture-reference/

You'd have 3 Git repositories: d2-fleet, d2-apps and d2-infra.
